### PR TITLE
feat(tile-converter): add EXT_feature_metadata extension support

### PR DESCRIPTION
**Pull request by dryabinin-actionengine**
Spec - https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_feature_metadata

### CODE REVIEW

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    case Boolean(extensions.EXT_feature_metadata): {
+      return handleExtFeatureMetadataExtension(attributes, extensions.EXT_feature_metadata, images);
+    }
+    case Boolean(extensions.EXT_mesh_features): {
+      // TODO Add support for EXT_mesh_features extension.
+      console.warn('EXT_mesh_features extension is not supported yet');
+      return [];
+    }
+    default: {
+      console.warn("Can't handle unsupported batchIds extensions");
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts:34]:
```suggestion
      console.warn("Can't handle unsupported extension ${extensionName}");
```

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+      devisorCounter -= 1;
+
+      if (devisorCounter === 0) {
+        currentValue++;
+        devisorCounter = divisor;
+      }
+    }
+  } else {
+    featureIds = Array<number>(featuresCount).fill(constant, 0, featuresCount);
```
Comment (ibgreen) [inline modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts:116]: Nit: Not sure how long these are, but maybe a typed array could be returned (e.g. Uint32Array) which would be more compact?

File: modules/tile-converter/test/i3s-converter/helpers/batch-ids-extensions.spec.js [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+  const images = [];
+
+  // @ts-ignore
+  const batchIds = handleBatchIdsExtensions(attributes, primitive, images);
+  const expectedResult = new Float32Array([0, 1, 2]);
+
+  t.deepEqual(batchIds, expectedResult);
+});
+
+test('#handleBatchIdsExtensions#handleExtFeatureMetadataExtension - Should return implicit batchIds for divisor = 1', async (t) => {
```
Comment (belom88) [inline modules/tile-converter/test/i3s-converter/helpers/batch-ids-extensions.spec.js:164]:
```suggestion
test('#handleBatchIdsExtensions#handleExtFeatureMetadataExtension - Should return implicit batchIds for divisor = 2', async (t) => {
```

**Review by belom88 — APPROVED** [review]
failing test

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts (hunk 2 above)
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts:116]: Will think how to improve it to use typed arrays but for now all batch ids functions chain waits for simple array type.

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  primitive: MeshPrimitive,
+  images: Image[]
+): number[] {
+  const extensions = primitive?.extensions;
+
+  if (!extensions) {
+    return [];
+  }
+
+  switch (true) {
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts]: Actually did as you said. It helps to get an access to extension name to handle it in default case.

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts (hunk 1 above)
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts]: Done

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+function handleExtFeatureMetadataExtension(
+  attributes: {
+    [key: string]: GLTFAccessorPostprocessed;
+  },
+  extFeatureMetadata: ExtFeatureMetadata,
+  images: Image[]
+): number[] {
+  // Take only first extension object to get batchIds attribute name.
+  const featureIdAttribute =
+    extFeatureMetadata?.featureIdAttributes && extFeatureMetadata?.featureIdAttributes[0];
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts]: Done

File: modules/tile-converter/test/i3s-converter/helpers/batch-ids-extensions.spec.js (hunk 3 above)
Comment (dryabinin-actionengine) [reply modules/tile-converter/test/i3s-converter/helpers/batch-ids-extensions.spec.js]: done

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  return batchIds;
+}
+
+/**
+ * Handle UVs if they are < 0 or > 1
+ * @param n
+ * @param m
+ */
+function emod(n: number, m: number): number {
+  return ((n % m) + m) % m;
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts]: Yes, I kept m as  only 1.

**Comment by ibgreen** [comment]
Were you able to look into this question:

> There are several existing files dedicated to the older batch id extension (I think they were not fully tested and probably inactive). Are those files independent of this, or replaced by this, or does would some additional alignment be required?

https://github.com/visgl/loaders.gl/tree/master/modules/3d-tiles/src/lib/classes

File: modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/tile-converter/src/i3s-converter/helpers/batch-ids-extensions.ts]: > This post made it to the front page of Hacker News and there were mixed thoughts about this pattern in the thread

If you read the [hackerrank link](https://news.ycombinator.com/item?id=26777090) in the header of your blog post, you can see that the programming community is rather divided around this pattern.

**Comment by dryabinin-actionengine** [comment]
> Were you able to look into this question:
> 
> > There are several existing files dedicated to the older batch id extension (I think they were not fully tested and probably inactive). Are those files independent of this, or replaced by this, or does would some additional alignment be required?
> 
> https://github.com/visgl/loaders.gl/tree/master/modules/3d-tiles/src/lib/classes



> Were you able to look into this question:
> 
> > There are several existing files dedicated to the older batch id extension (I think they were not fully tested and probably inactive). Are those files independent of this, or replaced by this, or does would some additional alignment be required?
> 
> https://github.com/visgl/loaders.gl/tree/master/modules/3d-tiles/src/lib/classes




My changes are independent from files which you shared with me. These files help to work with `InstancedModel3DTile` and `BatchedModel3DTile` formats. My changes related to parsing `Gltf3DTile`. Also `gltfLoader` already handled structured json data of  `EXT_feature_metadata` extension. What I need is to handle that extension from `gltfLoader` to get `batchIds` and convert them to I3S format. You can see difference between structures on picture below:

image

What we can possibly add `EXT_feature_metadata` extension with description to the `gltfLoader` documentation.

**Review by belom88 — APPROVED** [review]

File: modules/3d-tiles/src/types.ts [hunk 7]
```diff
@@ -88,3 +88,30 @@ type BufferView = {
   byteOffset: number;
   byteLength: number;
 };
+
+/**
+ * 3DTilesNext EXT_feature_metadata extension
+ * Spec - https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_feature_metadata
+ */
+export type ExtFeatureMetadata = {
```
[the remark this answers is not in the document]
Comment (dryabinin-actionengine) [reply modules/3d-tiles/src/types.ts]: Done

**Comment by ibgreen** [comment]
> image
> 

From my side it looks like it just a difference in packaging. These two extensions both seem to be containing batch ids. I understand that while these things are now packaged inside the gltf instead of in the tile, since that is more compact and convenient, I was mainly wondering if the underlying batch id data changed much. I still suspect we have two quite similar code paths for processing batch ids in the two cases. But it is of course possible that Cesium also changed the batch id data format so that code sharing no longer makes sense. 

No big deal, just wanted to explain my thinking.

> What we can possibly add `EXT_feature_metadata` extension with description to the `gltfLoader` documentation.

Good idea if we support gltf extensions we should list them.

**Pull request merged**